src/main/java/org/apache/mesos/scheduler/repair/RepairScheduler.java

```diff
@@ -0,0 +1,163 @@
+    private final LaunchConstrainer launchConstrainer;
+    private final AtomicReference<RepairStatus> repairStatusRef;
+
+    public RepairScheduler(
+            String targetConfigName,
+            StateStore stateStore,
+            TaskFailureListener failureListener,
+            RepairOfferRequirementProvider offerReqProvider,
+            OfferAccepter offerAccepter,
+            LaunchConstrainer launchConstrainer,
+            FailureMonitor failureMonitor,
+            AtomicReference<RepairStatus> repairStatusRef) {
+        this.targetConfigName = targetConfigName;
+        this.stateStore = stateStore;
+        this.offerReqProvider = offerReqProvider;
+        this.offerAccepter = offerAccepter;
+        this.failureMonitor = failureMonitor;
+        this.launchConstrainer = launchConstrainer;
+        this.repairStatusRef = repairStatusRef;
+        this.failureListener = failureListener;
+    }
+
+    public List<OfferID> resourceOffers(SchedulerDriver driver, List<Offer> offers, Block block)
+            throws Exception {
+        List<OfferID> acceptedOffers = new ArrayList<>();
+        List<TaskInfo> terminatedTasks = new ArrayList<>(getTerminatedTasks(block));
+
+        // Compute the stopped and failed pool for UI
+        RepairStatus oldStatus = repairStatusRef.get();
+        if (oldStatus == null) {
+            oldStatus = new RepairStatus(Collections.emptyList(), Collections.emptyList());
+        }
+        List<String> oldStopped = oldStatus.getStopped();
+        List<String> newStopped = new ArrayList<>(terminatedTasks.stream()
+                .filter(failureMonitor::hasFailed)
+                .map(TaskInfo::getName)
+                .collect(Collectors.toList()));
+        newStopped.addAll(oldStopped);
+        //TODO (ghartman) possibly write this list down in ZK, to ensure the UI doesn't change upon restart
+        newStopped = newStopped.stream().distinct().collect(Collectors.toList());
+        List<String> newFailed = terminatedTasks.stream()
+                .filter(it -> !failureMonitor.hasFailed(it))
+                .map(TaskInfo::getName)
+                .collect(Collectors.toList());
+
+        Optional<OfferRequirement> offerReq = Optional.empty();
+        boolean replacingMissing = false;
+
+        if (terminatedTasks.size() > 0) {
+            TaskInfo terminatedTask = terminatedTasks.get(new Random().nextInt(terminatedTasks.size()));
+            log.info("Found terminated task");
+            if (failureMonitor.hasFailed(terminatedTask)) {
+                log.info("Purging terminated task");
+                failureListener.taskFailed(terminatedTask.getTaskId());
+            } else {
+                log.info("Relaunching terminated task in place");
+                offerReq = Optional.of(offerReqProvider.getReplacementOfferRequirement(terminatedTask));
+            }
+        } else {
+            log.info("Launching a missing task");
```

Comment:
The concept of a missing Task has leaked out of the Kafka framework.  It is an implementation detail of the Kafka framework.  The work that needs to be done here is replacing permanently failed Tasks and restarting transiently failed Tasks.  "Missing" doesn't enter into the equation.